File: scripts/validate-schema.py
#!/usr/bin/env python3
"""
Validate skewer.yaml files against the JSON Schema.

Usage:
    python scripts/validate-schema.py skewer.yaml
    python scripts/validate-schema.py examples/*.yaml
"""

import sys
import json
from pathlib import Path

try:
    import yaml
    from jsonschema import validate, ValidationError, SchemaError
except ImportError:
    print("Error: Required packages not installed.", file=sys.stderr)
    print("Install with: pip install pyyaml jsonschema", file=sys.stderr)
    sys.exit(1)


def load_schema(schema_path: Path) -> dict:
    """Load the JSON schema."""
    with open(schema_path) as f:
        return json.load(f)


def load_yaml(yaml_path: Path) -> dict:
    """Load a YAML file."""
    with open(yaml_path) as f:
        return yaml.safe_load(f)


def validate_file(yaml_path: Path, schema: dict) -> tuple[bool, str]:
    """
    Validate a YAML file against the schema.

    Returns:
        Tuple of (success: bool, message: str)
    """
    try:
        data = load_yaml(yaml_path)
        validate(instance=data, schema=schema)
        return True, "✓ Valid"
    except ValidationError as e:
        # Format the validation error nicely
        path = " → ".join(str(p) for p in e.path) if e.path else "root"
        return False, f"✗ Validation error at {path}: {e.message}"
    except SchemaError as e:
        return False, f"✗ Schema error: {e.message}"
    except yaml.YAMLError as e:
        return False, f"✗ YAML parsing error: {e}"
    except Exception as e:
        return False, f"✗ Unexpected error: {e}"


def main():
    if len(sys.argv) < 2:
        print("Usage: validate-schema.py <yaml-file> [yaml-file ...]", file=sys.stderr)
        print("\nExample:", file=sys.stderr)
        print("  python scripts/validate-schema.py skewer.yaml", file=sys.stderr)
        print("  python scripts/validate-schema.py examples/*.yaml", file=sys.stderr)
        sys.exit(1)

    # Load schema
    script_dir = Path(__file__).parent
    schema_path = script_dir.parent / "skewer-schema.json"

    if not schema_path.exists():
        print(f"Error: Schema not found at {schema_path}", file=sys.stderr)
        sys.exit(1)

    try:
        schema = load_schema(schema_path)
    except Exception as e:
        print(f"Error loading schema: {e}", file=sys.stderr)
        sys.exit(1)

    # Validate each file
    yaml_files = [Path(arg) for arg in sys.argv[1:]]
    results = []

    for yaml_path in yaml_files:
        if not yaml_path.exists():
            results.append((yaml_path, False, f"✗ File not found: {yaml_path}"))
            continue

        success, message = validate_file(yaml_path, schema)
        results.append((yaml_path, success, message))

    # Print results
    max_path_len = max(len(str(p)) for p, _, _ in results)

    for yaml_path, success, message in results:
        path_str = str(yaml_path).ljust(max_path_len)
        print(f"{path_str}  {message}")

    # Exit with error if any validation failed
    if any(not success for _, success, _ in results):
        sys.exit(1)

    print(f"\n✓ All {len(results)} file(s) valid")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/validate-schema.go
// Validate skewer.yaml files against the JSON Schema using Go
//
// Build:
//   go build -o validate-schema scripts/validate-schema.go
//
// Usage:
//   ./validate-schema skewer.yaml
//   ./validate-schema examples/*.yaml
//
// Dependencies:
//   go get github.com/xeipuuv/gojsonschema
//   go get gopkg.in/yaml.v3

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/xeipuuv/gojsonschema"
	"gopkg.in/yaml.v3"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: validate-schema <yaml-file> [yaml-file ...]")
		fmt.Fprintln(os.Stderr, "")
		fmt.Fprintln(os.Stderr, "Example:")
		fmt.Fprintln(os.Stderr, "  ./validate-schema skewer.yaml")
		fmt.Fprintln(os.Stderr, "  ./validate-schema examples/*.yaml")
		os.Exit(1)
	}

	// Load schema
	exePath, err := os.Executable()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting executable path: %v\n", err)
		os.Exit(1)
	}

	scriptDir := filepath.Dir(exePath)
	schemaPath := filepath.Join(scriptDir, "..", "skewer-schema.json")

	// Try current directory if not found
	if _, err := os.Stat(schemaPath); os.IsNotExist(err) {
		schemaPath = "skewer-schema.json"
	}

	schemaLoader := gojsonschema.NewReferenceLoader("file://" + schemaPath)

	// Validate each file
	results := []result{}
	maxPathLen := 0

	for _, yamlPath := range os.Args[1:] {
		if len(yamlPath) > maxPathLen {
			maxPathLen = len(yamlPath)
		}

		success, message := validateFile(yamlPath, schemaLoader)
		results = append(results, result{
			path:    yamlPath,
			success: success,
			message: message,
		})
	}

	// Print results
	for _, r := range results {
		pathStr := fmt.Sprintf("%-*s", maxPathLen, r.path)
		fmt.Printf("%s  %s\n", pathStr, r.message)
	}

	// Exit with error if any validation failed
	failed := false
	for _, r := range results {
		if !r.success {
			failed = true
			break
		}
	}

	if failed {
		os.Exit(1)
	}

	fmt.Printf("\n✓ All %d file(s) valid\n", len(results))
	os.Exit(0)
}

type result struct {
	path    string
	success bool
	message string
}

func validateFile(yamlPath string, schemaLoader gojsonschema.JSONLoader) (bool, string) {
	// Check file exists
	if _, err := os.Stat(yamlPath); os.IsNotExist(err) {
		return false, fmt.Sprintf("✗ File not found: %s", yamlPath)
	}

	// Load YAML file
	data, err := os.ReadFile(yamlPath)
	if err != nil {
		return false, fmt.Sprintf("✗ Error reading file: %v", err)
	}

	// Parse YAML
	var yamlData interface{}
	if err := yaml.Unmarshal(data, &yamlData); err != nil {
		return false, fmt.Sprintf("✗ YAML parsing error: %v", err)
	}

	// Convert to JSON (required for gojsonschema)
	jsonData, err := json.Marshal(yamlData)
	if err != nil {
		return false, fmt.Sprintf("✗ Error converting to JSON: %v", err)
	}

	// Validate against schema
	documentLoader := gojsonschema.NewBytesLoader(jsonData)
	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
	if err != nil {
		return false, fmt.Sprintf("✗ Validation error: %v", err)
	}

	if !result.Valid() {
		// Format the first error nicely
		if len(result.Errors()) > 0 {
			err := result.Errors()[0]
			path := err.Field()
			if path == "(root)" {
				path = "root"
			}
			return false, fmt.Sprintf("✗ Validation error at %s: %s", path, err.Description())
		}
		return false, "✗ Validation failed"
	}

	return true, "✓ Valid"
}

func formatPath(path string) string {
	// Convert "(root).sites.west.platform" to "sites → west → platform"
	path = strings.TrimPrefix(path, "(root)")
	path = strings.TrimPrefix(path, ".")
	parts := strings.Split(path, ".")
	return strings.Join(parts, " → ")
}
